nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/NiFiRegistryMode.groovy

```diff
@@ -0,0 +1,383 @@
+
+        // Options for input/output identity-providers.xml files
+        cli.i(longOpt: 'identityProvidersXml',
+                args: 1,
+                argName: 'file',
+                'The identity-providers.xml file containing unprotected config values, overwritten if no output file specified.')
+        cli.I(longOpt: 'outputIdentityProvidersXml',
+                args: 1,
+                argName: 'file',
+                'The destination identity-providers.xml file containing protected config values.')
+
+        return cli
+
+    }
+
+    class Configuration {
+
+        OptionAccessor rawOptions
+
+        boolean usingRawKeyHex
+        boolean usingPassword
+        boolean usingBootstrapKey
+
+        String encryptionKey
+        String decryptionKey
+
+        SensitivePropertyProvider encryptionProvider
+        SensitivePropertyProvider decryptionProvider
+
+        boolean writingKeyToBootstrap = false
+        String inputBootstrapPath
+        String outputBootstrapPath
+
+        boolean handlingNiFiRegistryProperties = false
+        String inputNiFiRegistryPropertiesPath
+        String outputNiFiRegistryPropertiesPath
+        NiFiRegistryPropertiesEncryptor propertiesEncryptor
+
+        boolean handlingIdentityProviders = false
+        String inputIdentityProvidersPath
+        String outputIdentityProvidersPath
+        NiFiRegistryIdentityProvidersXmlEncryptor identityProvidersXmlEncryptor
+
+        boolean handlingAuthorizers = false
+        String inputAuthorizersPath
+        String outputAuthorizersPath
+        NiFiRegistryAuthorizersXmlEncryptor authorizersXmlEncryptor
+
+        Configuration() {
+        }
+
+        Configuration(OptionAccessor options) {
+            this.rawOptions = options
+
+            validateOptions()
+
+            // Set input bootstrap.conf path
+            inputBootstrapPath = rawOptions.b
+
+            // Determine key for encryption (required)
+            determineEncryptionKey()
+            if (!encryptionKey) {
+                throw new RuntimeException("Failed to configure tool, could not determine encryption key. Must provide -p, -k, or -b. If using -b, bootstrap.conf argument must already contain master key.")
+            }
+            encryptionProvider = new AESSensitivePropertyProvider(encryptionKey)
+
+            // Determine key for decryption (if migrating)
+            determineDecryptionKey()
+            if (!decryptionKey) {
+                logger.debug("No decryption key specified via options, so if any input files require decryption prior to re-encryption (i.e., migration), this tool will fail.")
+            }
+            decryptionProvider = decryptionKey ? new AESSensitivePropertyProvider(decryptionKey) : null
+
+            writingKeyToBootstrap = (usingPassword || usingRawKeyHex || rawOptions.B)
+            if (writingKeyToBootstrap) {
+                outputBootstrapPath = rawOptions.B ? rawOptions.B : inputBootstrapPath
```

Comment:
In the future, the Elvis operator condenses this in Groovy (not necessary to make the change for this PR). 

`outputBootstrapPath = rawOptions.B ?: inputBootstrapPath`

Reply:
Oh cool, I didn't know about that Groovy idiom - thanks!

- [ ] (optional, time permitting) change to using Elvis operator where applicable